File: code/ch08-01/game.erl
%% @author J D Eisenberg <jdavid.eisenberg@gmail.com>
%% @doc Play the card game "war" with two players.
%% @copyright 2013 J D Eisenberg
%% @version 0.1

-module(game).
-export([play_game/0, dealer/0, player/2, value/1]).

%% @doc create a dealer
play_game() ->
  spawn(game, dealer, []).

dealer() ->
  random:seed(now()),
  DealerPid = self(),
  Deck = cards:shuffle(cards:make_deck()),
  {P1Cards, P2Cards} = lists:split(trunc(length(Deck) / 2), Deck),
  io:format("About to spawn players each with ~p cards.~n",
    [trunc(length(Deck) / 2)]),
  P1 = spawn(game, player, [DealerPid, P1Cards]),
  P2 = spawn(game, player, [DealerPid, P2Cards]),
  io:format("Spawned players ~p and ~p~n", [P1, P2]),
  dealer([P1, P2], pre_battle, [], [], 0, []).

%% The dealer has to keep track of the players' process IDs,
%% the cards they have given to the dealer for comparison,
%% how many players have responded (0, 1, or 2), and the pile
%% in the middle of the table in case of a war.

dealer(Pids, State, P1Cards, P2Cards, Count, Pile) ->
  [P1, P2] = Pids,
  NCards = if
    Pile == []  -> 1;
    Pile /= [] -> 3
  end,  
  case State of
    pre_battle ->
      P1 ! {give_cards, NCards},
      P2 ! {give_cards, NCards},
      dealer(Pids, await_battle, P1Cards, P2Cards, Count, Pile);
    await_battle ->
      receive
        {accept, Pid, Data} ->
          NextCount = Count + 1,
          case Pid of
            P1 -> Next_P1Cards = Data, Next_P2Cards = P2Cards;
            P2 -> Next_P1Cards = P1Cards, Next_P2Cards = Data
          end
      end,
      if
        NextCount == 2 -> NextState = check_cards;
        NextCount /= 2 -> NextState = State
      end,
      dealer(Pids, NextState, Next_P1Cards, Next_P2Cards,
        NextCount, Pile);
    check_cards ->
      Winner = game_winner(P1Cards, P2Cards),
      case Winner of
        0 ->
          io:format("Compare ~p to ~p~n", [P1Cards, P2Cards]),
          NewPile = Pile ++ P1Cards ++ P2Cards,
          case battle_winner(P1Cards, P2Cards) of
             0 -> dealer(Pids, pre_battle, [], [], 0, NewPile);
             1 ->
              P1 ! {take_cards, NewPile},
              dealer(Pids, await_confirmation, [], [], 0, []);
             2 ->
              P2 ! {take_cards, NewPile},
              dealer(Pids, await_confirmation, [], [], 0, [])
           end;
        3 ->
          io:format("It's a draw!~n"),
          end_game(Pids);
        _ ->
          io:format("Player ~p wins~n", [Winner]),
          end_game(Pids)
      end;
    await_war->
      io:format("Awaiting war~n");
    await_confirmation ->
      io:format("Awaiting confirmation of player receiving cards~n"),
      receive
        {confirmed, _Pid, _Data} ->
        dealer(Pids, pre_battle, [], [], 0, [])
      end
  end.

end_game(Pids) ->
  lists:foreach(fun(Process) -> exit(Process, kill) end, Pids),
  io:format("Game finished.~n").

%% Do we have a winner? If both players are out of cards,
%% it's a draw. If one player is out of cards, the other is the winner.

game_winner([], []) -> 3;
game_winner([], _) -> 2;
game_winner(_, []) -> 1;
game_winner(_, _) -> 0.

battle_winner(P1Cards, P2Cards) ->
  V1 = value(hd(lists:reverse(P1Cards))),
  V2 = value(hd(lists:reverse(P2Cards))),
  Winner = if
    V1 > V2 -> 1;
    V2 > V1 -> 2;
    V1 == V2 -> 0
  end,
  io:format("Winner of ~p vs. ~p is ~p~n", [V1, V2, Winner]),
  Winner = Winner.

player(Dealer, Hand) ->
  receive
    {Command, Data} ->
      case Command of
        give_cards ->
          {ToSend, NewHand} = give_cards(Hand, Data),
          io:format("Sending ~p to ~p~n", [ToSend, Dealer]),
          Dealer!{accept, self(), ToSend};
        take_cards ->
          io:format("~p now has ~p (cards)~n", [self(),
            length(Data) + length(Hand)]),
          NewHand = Hand ++ Data,
          Dealer!{confirmed, self(), []}
      end
  end,
  player(Dealer, NewHand).

%% Player gives N cards from current Hand. N is 1 or 3,
%% depending if there is a war or not.
%% If a player is asked for 3 cards but doesn't have enough,
%% give all the cards in the hand.
%% This function returns a tuple: {[cards to send], [remaining cards in hand]}

give_cards([], _N) -> {[],[]};
give_cards([A], _N) -> {[A],[]};
give_cards([A, B], N) ->
  if
    N == 1 -> {[A], [B]};
    N == 3 -> {[A, B], []}
  end;
give_cards(Hand, N) ->
  if
    N == 1 -> {[hd(Hand)], tl(Hand)};
    N == 3 ->
      [A, B, C | Remainder] = Hand,
      {[A, B, C], Remainder}
  end.
  
%% @doc Returns the value of a card. Aces are high; K > Q > J
-spec(value({cards:card()}) -> integer()).

value({V, _Suit}) ->
  if
    is_integer(V) -> V;
    is_list(V) ->
      case hd(V) of
        $J -> 11;
        $Q -> 12;
        $K -> 13;
        $A -> 14
      end
  end.

File: code/ch08-01/game.ex
defmodule Game do

  def start() do
    deck = Cards.shuffle(Cards.make_deck(
#      ["A", 2, 3, 4, 5, 6, 7, 8, 9, 10, "J", "Q", "K"],
      [2, 3, 4, 5],
      ["Clubs", "Diamonds", "Hearts", "Spades"]))
    {hand1, hand2} = Enum.split(deck, trunc(Enum.count(deck) / 2))
    player1 = spawn(Player, :start, [hand1])
    player2 = spawn(Player, :start, [hand2])
    play([player1, player2], :pre_battle, [], [], 0, [])
  end

  @doc """
  Arguments are: list of player pids, state of game (atom),
  cards received from player 1, cards received from player 2,
  number of players who have given cards, and pile in middle of table
  """
  @spec play(list, atom, list, list, integer, list) :: nil
  
  def play(players, state, cards1, cards2, n_received,
    pile) do
    [player1, player2] = players
    case state do
    
      # Ask players to give you cards. If there are no
      # cards in the pile, it's a simple battle; otherwise
      # it's a war.
      
      :pre_battle ->
        IO.puts("") # for spacing
        case pile do
          [] ->
            IO.puts("Requesting 1 card from each player")
            request_cards(players, 1)
          _ ->
            IO.puts("Requesting 3 cards from each player")
            request_cards(players, 3)
        end
        play(players, :await_battle, cards1, cards2, n_received, pile)
 
      # When both players have given you their card(s),
      # you need to check them.
      
      :await_battle when n_received == 2 ->
        play(players, :check_cards, cards1, cards2, 0, pile)
      
      # Otherwise, wait for players to send you card(s). Each time
      # you receive card(s), remember them and add one to the
      # total number of players who have responded.
      
      :await_battle ->
        receive do
          {:take, new_cards, from} ->
            IO.puts("Got #{inspect(new_cards)} from #{inspect(from)}")
            cond do
              from == player1 -> play(players, state, new_cards,
                cards2, n_received + 1, pile)
              from == player2 -> play(players, state, cards1,
                new_cards, n_received + 1, pile)
            end
        end
      
      # If both people have run out of cards, it's a draw.
      # If one person is out of cards, the other player is the winner.
      # Otherwise, evaluate the cards and prepare for next
      # battle or war.
      
      :check_cards ->
        cond do
          cards1 == [] and cards2 == [] ->
            IO.puts("Draw")
            endgame(players)
          cards1 == [] ->
            IO.puts("Player 2 wins")
            endgame(players)
          cards2 == [] ->
            IO.puts("Player 1 wins")
            endgame(players)
          true -> 
            new_pile = evaluate(players, cards1, cards2, pile)
            play(players, :pre_battle, [], [], 0, new_pile)
        end   
    end
  end
  
  @spec evaluate(list, list, list, list) :: list
  
  # Evaluate the cards from both players. If their
  # values match, add them to the pile and.
  # If they don't, the winner is told to pick up the cards
  # (and whatever's in the pile), and the pile is cleared.
  #
  # Wait for players to respond before proceeding with the
  # game. Otherwise, a player with an empty hand might be
  # asked to give a card before picking up the cards she won.
  
  defp evaluate(players, cards1, cards2, pile) do
    [player1, player2] = players
    v1 = card_value(hd(cards1))
    v2 = card_value(hd(cards2))
    IO.puts("Value of card 1 is #{v1}; value of card 2 is #{v2}")
    new_pile = List.concat([pile, cards1, cards2])
    IO.puts("Card pile is now #{inspect(new_pile)}")
    cond do
      v1 == v2 ->
        IO.puts("Equal values; going to war.")
        new_pile  # it's a war
      v1 > v2 ->
        IO.puts("Telling player 1 to pick up the cards")
        player1 <- {:pick_up, new_pile, self()}
        wait_for_pickup()
        []
      v2 > v1 ->
        IO.puts("Telling player 2 to pick up the cards")
        player2 <- {:pick_up, new_pile, self()}
        wait_for_pickup()
        []
    end
  end
  
  # Wait for player to pick up cards
  @spec wait_for_pickup() :: pid
  def wait_for_pickup() do
    receive do
      {:got_cards, player} ->
        IO.puts("Player #{inspect(player)} picked up cards.")
        player
    end
  end
  
  # Send each player a requst to send n cards
  @spec request_cards(list, integer) :: nil
  
  defp request_cards([p1, p2], n) do
    p1 <- {:give, n, self()}
    p2 <- {:give, n, self()}
  end
  
  # Send message to all players to exit their receive loop.
  @spec endgame(list) :: nil
  
  defp endgame(players) do
    Enum.each(players, fn(x) -> x <- :game_over end)
  end
  
  # Return the value of a card; Aces are high.
  @spec card_value(tuple) :: integer
  
  defp card_value({value, _suit}) do
    case value do
      "A" -> 14
      "K" -> 13
      "Q" -> 12
      "J" -> 11
      _ -> value
    end
  end
end
